docs/proposals/volume-provisioning.md

```diff
@@ -0,0 +1,509 @@
+    actual attribute.
+
+4.  `PersistentVolume` instances should be labelled with configuration
+    parameters that were used for their creation, such as availability zone or
+    storage type (SSD/HDD). This should be done in an admission controller.
+
+5.  `PersistentVolumeClaim.Spec.Class` attribute is added to claims. This
+    attribute specifies that only a volume with equal
+    `PersistentVolume.Spec.Class` value can satisfy a claim.
+
+    During alpha and beta development, `Class` is just an annotation and not
+    actual attribute.
+
+6.  When requesting a volume with certain configuration,
+    `PersistentVolumeClaim.Spec.Selector` should be set in a way that it matches
+    only volumes with this configuration (see point 3.)
+    When both `PersistentVolumeClaim.Spec.Selector` and
+    `PersistentVolumeClaim.Spec.Class` is set, resulting volume must match both
+    selectors in the claim and in the storage class.
+
+7.  The persistent volume controller be modified to dynamically provision
+    volumes using storage classes and claim's `Selector` to create a volume that
+    has properties requested by both `PersistentVolumeClaim.Spec.Class` and
+    `PersistentVolumeClaim.Spec.Selector`.
+
+8.  The existing provisioner plugin implementations be modified to allow
+    parameterization as appropriate via storage classes and claim selectors.
+
+9.  The existing alpha dynamic provisioning feature be phased out in the
+    next release.
+
+### Controller workflow for provisioning volumes
+
+1.  When a new claim is submitted, the controller attempts to find an existing
+    volume that will fulfill the claim.
+
+    1. If the claim has non-empty `claim.Spec.Class`, only PVs with the same
+        `pv.Spec.Class` are considered.
+
+    2. If the claim has empty `claim.Spec.Class`, all existing PVs are
+        considered.
+
+    All "considered" volumes are evaluated with `claim.Spec.Selector` and the
+    smallest matching one is then bound to the claim.
+
+2.  If no volume is found for the claim and `claim.Spec.Class` is not set or is
+    empty string, the controller will use a default StorageClass specified on
+    kube-controller-manager command line. If no default StorageClass is
+    configured, dynamic provisioning is disabled and the controller goes back to
+    step 1. and periodically retries finding a matching volume or storage class
+    again until a match is found. The claim is `Pending` during this period.
+
+3.  Either with `claim.Spec.Class` or the default storage class name, the
+    controller tries to find instance of StorageClass with this name.  If no
+    such StorageClass is found, the controller goes back to step 1. and
+    periodically retries finding a matching volume or storage class again until
+    a match is found. The claim is `Pending` during this period.
+
+4.  With StorageClass instance, the controller finds volume plugin specified by
+    StorageClass.ProvisionerType.
```

Comment:
I thought that `ProvisionerType` is just name of the storage plugin that will handle provisioning, was there any other interpretations?